packages/sdk/src/search/useSearchState.ts

```diff
@@ -0,0 +1,226 @@
+      type: 'removeFacet'
+      payload: Facet
+    }
+  | {
+      type: 'toggleFacet'
+      payload: Facet
+    }
+  | {
+      type: 'toggleFacets'
+      payload: Facet[]
+    }
+
+const removeFacet = (state: State, facet: Facet): State => {
+  const { value } = facet
+
+  const index = state.selectedFacets.findIndex((x) => x.value === value)
+
+  if (index < 0) {
+    throw new SDKError(`Cannot remove ${value} from search params`)
+  }
+
+  // We can't allow removing the first facet, otherwise we would loose
+  // the navigation context
+  //
+  // TODO: Remove returning the base selected facets to the frontend since
+  // we won't be unselecting it anyways
+  return {
+    ...state,
+    selectedFacets: state.selectedFacets.filter(
+      (_, it) => it === 0 || it !== index
+    ),
+  }
+}
+
+export const setFacet = (
+  state: State,
+  facet: Facet,
+  unique: boolean
+): State => {
+  if (unique === true) {
+    const index = state.selectedFacets.findIndex((f) => f.key === facet.key)
+
+    if (index > -1) {
+      return {
+        ...state,
+        selectedFacets: state.selectedFacets.map((f, it) =>
+          it === index ? facet : f
+        ),
+      }
+    }
+  }
+
+  return {
+    ...state,
+    selectedFacets: [...state.selectedFacets, facet],
+  }
+}
+
+const toggleFacet = (state: State, item: Facet) => {
+  const found = state.selectedFacets.find(
+    (facet) => facet.key === item.key && facet.value === item.value
+  )
+
+  if (found !== undefined) {
+    return removeFacet(state, item)
+  }
+
+  return setFacet(state, item, false)
+}
+
+const toggleFacets = (state: State, items: Facet[]) =>
+  items.reduce((item, s) => toggleFacet(item, s), state)
+
+export const reducer = (state: State, action: Action) => {
+  switch (action.type) {
+    case 'setSort':
+      if (!isSearchSort(action.payload)) {
+        throw new SDKError(`Sort param ${action.payload} is unknown`)
+      }
+
+      return state.sort === action.payload
+        ? state
+        : {
+            ...state,
+            sort: action.payload,
+          }
+
+    case 'setTerm':
+      return state.term === action.payload
+        ? state
+        : {
+            ...state,
+            term: action.payload,
+          }
+
+    case 'setPage':
+      return state.page === action.payload
+        ? state
+        : {
+            ...state,
+            page: action.payload,
+          }
+
+    case 'setFacet':
+      return setFacet(state, action.payload.facet, action.payload.unique)
+
+    case 'removeFacet':
+      return removeFacet(state, action.payload)
+
+    case 'toggleFacet':
+      return toggleFacet(state, action.payload)
+
+    case 'toggleFacets':
+      return toggleFacets(state, action.payload)
+
+    default:
+      throw new SDKError(`Uknown action of search state machine`)
+  }
+}
+
+export const useSearchState = (initialState: Partial<State>) => {
+  const [state, dispatch] = useReducer(reducer, initialState, initialize)
+
+  return {
```

Comment:
Shouldn't this return be inside the memo?

Reply:
I think you are right. I'll move the useMemo from `Provider.tsx` to here